rust/xnvme/src/device.rs

```diff
@@ -0,0 +1,374 @@
+pub use xnvme_sys::*;
+pub type size_t = ::std::os::raw::c_ulong;
+use super::Result;
+
+/*
+ * Creates a device handle (::xnvme_dev) based on the given device-uri
+ *
+ * @param dev_uri File path "/dev/nvme0n1" or "0000:04.01"
+ * @param opts Options for library backend and system-interfaces
+ *
+ * @return On success, a handle to the device. On error, NULL is returned and
+ * `errno` set to indicate the error.
+ */
+
+#[derive(Debug)]
+pub struct Device {
+    // Invariant: `device` must always point to a valid `xnvme_dev`.
+    pub device: *mut xnvme_sys::xnvme_dev,
+}
+
+impl Device {
+    pub unsafe fn new(device: *mut xnvme_sys::xnvme_dev) -> Self {
+        Self { device }
+    }
+
+    pub fn as_raw(&self) -> *mut xnvme_sys::xnvme_dev {
+        self.device
+    }
+
+    pub fn open(device_uri: &str, options: &crate::options::Options) ->
+                                                                Result<Self> {
+        let device_uri_c = std::ffi::CString::new(device_uri)?;
+
+        let dev = unsafe { xnvme_sys::xnvme_dev_open(
+                                        device_uri_c.as_ptr(),
+                                        &mut options.as_raw())
+                            };
+
+        // Todo: check non null
+
+        // SAFETY: This invocation complies with callee safety requirement
+        // because we just constructed dev above and it is therefore a valid
+        // instance
+
+        let device = unsafe { Self::new(dev) };
+        Ok(device)
+    }
+}
+
+/*
+ * Returns the geometry of the given device
+ *
+ * @param dev Device handle obtained with xnvme_dev_open()
+ *
+ * @return The geometry (struct xnvme_geo) of given device handle
+ */
+pub fn dev_get_geo(dev: *const xnvme_dev) -> *const xnvme_geo {
```

Comment:
The rest of the functions in this file seem to be rather pointless. What is the purpose of these? The documentation is a copy of the C documentation which is already rendered as part of `xnvme_sys`.